File: 2024/src/day1/day1main.go
package day1

import (
	"regexp"
	"slices"
	"strconv"
)

func Day1Pt2(inputLines []string) int {
	// list1, list2 := parseInput("day1_test1.txt")
	// list1, list2 := parseInput("day1.txt")
	list1, list2 := parseInput(inputLines)
	scores := []int{}
	for _, l := range list1 {
		rCount := 0
		for _, r := range list2 {
			if l == r {
				rCount++
			}
		}
		score := l * rCount
		scores = append(scores, score)
	}
	scoreSum := 0
	for _, score := range scores {
		scoreSum += score
	}
	return scoreSum
}

func Day1Pt1(inputLines []string) int {
	// list1, list2 := parseInput("day1_test1.txt")
	// list1, list2 := parseInput("day1.txt")
	list1, list2 := parseInput(inputLines)
	cmpFn := func(i int, j int) int {
		return i - j
	}
	slices.SortStableFunc(list1, cmpFn)
	slices.SortStableFunc(list2, cmpFn)

	listDiffs := []int{}
	for i := range list1 {
		diff := list1[i] - list2[i]
		if diff < 0 {
			diff = -diff
		}
		listDiffs = append(listDiffs, diff)
	}
	diffSum := 0
	for _, n := range listDiffs {
		diffSum += n
	}
	return diffSum
}

func parseInput(inputLines []string) ([]int, []int) {
	list1 := []int{}
	list2 := []int{}
	rx := regexp.MustCompile(`^(\d+)\s+(\d+)$`)
	for _, currLine := range inputLines {
		rxMatch := rx.FindStringSubmatch(currLine)
		n1, err := strconv.Atoi(rxMatch[1])
		if err != nil {
			panic(err)
		}
		n2, err := strconv.Atoi(rxMatch[2])
		if err != nil {
			panic(err)
		}
		list1 = append(list1, n1)
		list2 = append(list2, n2)
	}
	return list1, list2
}

File: 2024/src/day1/day1-main.lua

local function parseInput(inputLines)
  local list1 = {}
  local list2 = {}
  for _, currLine in ipairs(inputLines) do
    local val1, val2 = currLine:gmatch("(%d+)%s+(%d+)")()
    local n1 = tonumber(val1)
    local n2 = tonumber(val2)
    table.insert(list1, n1)
    table.insert(list2, n2)
  end
  local res = {
    list1 = list1,
    list2 = list2,
  }
  return res
end

local function day2Pt2(inputLines)
  local parsedInput = parseInput(inputLines)
  local list1 = parsedInput.list1
  local list2 = parsedInput.list2
  local scores = {}
  for _, l in ipairs(list1) do
    local rCount = 0
    for _, r in ipairs(list2) do
      if l == r then
        rCount = rCount + 1
      end
    end
    local score = l * rCount
    table.insert(scores, score)
  end
  local scoreSum = 0
  for _, score in ipairs(scores) do
    scoreSum = scoreSum + score
  end
  return scoreSum
end

--[[ 
  2196996 - correct
]]
local function day1Pt1(inputLines)
  local parsedInput = parseInput(inputLines)
  local list1 = parsedInput.list1
  local list2 = parsedInput.list2
  table.sort(list1)
  table.sort(list2)
  local listDiffs = {}
  for i, n1 in ipairs(list1) do
    local currDiff = math.abs(n1 - list2[i])
    table.insert(listDiffs, currDiff)
  end
  local totalDistance = 0
  for _, n in ipairs(listDiffs) do
    totalDistance = totalDistance + n
  end
  return totalDistance
end

local day1Module = {
  day1Pt1 = day1Pt1,
  day2Pt2 = day2Pt2,
}

return day1Module
